doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,191 @@
+
+❯ cat project-info.json| jq | head
+{
+  "project": {
+    "artifactId": "mvn2nix-maven-plugin",
+    "groupId": "org.nixos.mvn2nix",
+    "version": "1.2.2",
+    "classifier": "",
+    "extension": "maven-plugin",
+    "dependencies": [
+      {
+        "artifactId": "javax.json-api",
+
+```
+
+This file is then given to the _buildMaven_ function, and it returns 3 properties.
+
+**repo**:
+    A Maven repository that is a symlink farm of all the dependencies found in the _project-info.json_
+
+
+**build**:
+    A simple derivation that runs through _mvn compile_ & _mvn package_ to build the JAR. You may use this as inspiration for more complicated derivations.
+
+Here is an example of building the Maven repository
+```nix
+{ pkgs ? import <nixpkgs> {} }:
+with pkgs;
+let build = buildMaven ./project-info.json;
+in
+build.repo
+```
+
+The benefit over the _double invocation_, is that the _/nix/store_ entry is a _linkFarm_ of every package, so that changes to your dependency set doesn't involve downloading everything from scratch.
+
+```bash
+❯ tree /nix/store/5zg4jxj7897lcasd7zpvjrk8drnmiaf5-maven-repository | head
+
+/nix/store/5zg4jxj7897lcasd7zpvjrk8drnmiaf5-maven-repository
+├── antlr
+│   └── antlr
+│       └── 2.7.2
+│           ├── antlr-2.7.2.jar -> /nix/store/d027c8f2cnmj5yrynpbq2s6wmc9cb559-antlr-2.7.2.jar
+│           └── antlr-2.7.2.pom -> /nix/store/mv42fc5gizl8h5g5vpywz1nfiynmzgp2-antlr-2.7.2.pom
+├── aopalliance
+│   └── aopalliance
+│       └── 1.0
+│           ├── aopalliance-1.0.jar -> /nix/store/yy66cplfasm9m5r61hi09vbk4mkk47rl-aopalliance-1.0.jar
+```
+
+## Runnable JAR
+
+You are able to run `mvn compile` but how do you get a runnable artifact without having to use something like building an UberJar ?
+
+We will use the same repositories we built above (either _double invocation_ or _buildMaven_) to setup a classpath for our Jar.
+
+Setup your _pom.xml_ to create a Jar with the following manifest:
+```xml
+<plugin>
+    <artifactId>maven-jar-plugin</artifactId>
+    <version>3.2.0</version>
+    <configuration>
+        <archive>
+            <manifest>
+                <addClasspath>true</addClasspath>
+                <classpathPrefix>lib/</classpathPrefix>
+                <classpathLayoutType>repository</classpathLayoutType>
+                <mainClass>$YOUR_MAIN_CLASS_HERE</mainClass>
+            </manifest>
+            <manifestEntries>
+                <Class-Path>.</Class-Path>
+            </manifestEntries>
+        </archive>
+    </configuration>
+</plugin>
+```
+
+> This instructs the Jar to look for the necessary dependencies in the _lib/_ relative folder. The layout of the folder is also in the _maven repository_ style.
+
+```nix
+# create a repository using one of the methods above
+let repository = ....;
+in
+stdenv.mkDerivation rec {
+  pname = "your-pkg";
+  version = "0.1";
+  name = "${pname}-${version}";
+  src = ./.;
+  buildInputs = [ jdk maven makeWrapper ];
```

Comment:
Technically they aren't needed as they are referenced thanks to the `makeWrapper` call.

```suggestion
  buildInputs = [ makeWrapper ];
```

Reply:
We're invoking `mvn` in via `PATH` in `buildPhase` though. I figure a `jdk` command is also necessary for `mvn compile`

Reply:
> I figure a jdk command is also necessary for mvn compile

`maven` closes `pkgs.jdk` in `makeWrapper` (https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/apache-maven/builder.sh) which sometimes is wrong Java version (8 vs.11 issue) for a given build.

Reply:
I'll add a section on "Choosing the correct JDK" 
The current **mvn** derivation allows overriding _JAVA_HOME_

```bash
❯ cat /nix/store/2hhmmj0vbb5d181nfx2mx3p7k54q44ij-apache-maven-3.6.3/bin/mvn

#! /nix/store/6737cq9nvp4k5r70qcgf61004r0l2g3v-bash-4.4-p23/bin/bash -e
export JAVA_HOME=${JAVA_HOME-'/nix/store/knh4n8fzrslyzgrw6p8nh5v07s0ycznc-openjdk-8u265-ga'}
exec "/nix/store/2hhmmj0vbb5d181nfx2mx3p7k54q44ij-apache-maven-3.6.3/maven/bin/mvn"  "$@"
```

What i've done in the past is just make sure to set JAVA_HOME to the correct JDK (i.e. JDK8 vs JDK11).
An alternative approach would also be to `overrideAttrs`